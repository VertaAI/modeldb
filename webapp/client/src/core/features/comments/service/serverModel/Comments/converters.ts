import { IComment } from '../../../Model';

export const convertServerComment = <Comment extends IComment>(
  serverComment: any
): Comment => {
  return {
    id: serverComment.id,
    dateTime: new Date(Number(serverComment.date_time)),
    message: serverComment.message,
  } as Comment;
};
