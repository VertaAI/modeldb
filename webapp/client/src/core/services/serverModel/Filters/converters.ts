import { IFilterData } from 'core/features/filter/Model';
import { getServerFilterValueType, getServerFilterOperator } from './Filters';

export interface IFilterConverter<
  R extends IServerFiltersInRequest,
  T extends IFilterData = IFilterData
> {
  predicate: (filter: IFilterData) => filter is T;
  convert: (filter: T, request: R) => Promise<R>;
}
export interface IServerFiltersInRequest {
  predicates?: any[];
}
export const makeFilterConverter = <
  R extends IServerFiltersInRequest,
  T extends IFilterData = IFilterData
>(
  filterConverter: IFilterConverter<R, T>
) => filterConverter;

const defaultFilterConverter = makeFilterConverter<IServerFiltersInRequest>({
  predicate: (_): _ is IFilterData => true,
  convert: (filter, request) => {
    return Promise.resolve({
      ...request,
      predicates: (request.predicates || []).concat({
        key: filter.name.toLocaleLowerCase() === 'tag' ? 'tags' : filter.name,
        value: Array.isArray(filter.value) ? filter.value[0] : filter.value,
        value_type: getServerFilterValueType(filter.type),
        operator: getServerFilterOperator(filter),
      }),
    });
  },
});

export const makeAddFiltersToRequest: <R extends IServerFiltersInRequest>(
  additionalFilterConverters?: Array<IFilterConverter<R, any>>
) => (filters: IFilterData[]) => (request: R) => Promise<R> = (
  additionalFilterConverters = []
) => filters => request => {
  const filterConverters = [
    ...additionalFilterConverters,
    defaultFilterConverter,
  ];

  return filters
    .map(filter => {
      const appropriateFilterConverter = filterConverters.find(
        ({ predicate }) => predicate(filter)
      );
      if (appropriateFilterConverter) {
        return [filter, appropriateFilterConverter.convert] as [
          IFilterData,
          IFilterConverter<any>['convert']
        ];
      }
      throw new Error(`not found converter for filter = ${filter}`);
    })
    .reduce((resPromise, [filter, convert]) => {
      return resPromise.then(resRequest => convert(filter, resRequest));
    }, Promise.resolve(request));
};
