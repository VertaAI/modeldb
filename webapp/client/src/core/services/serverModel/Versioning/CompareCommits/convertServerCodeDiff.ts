import {
  IGitCodeDiff,
  ICodeBlobDiff,
  INotebookCodeDiff,
} from 'core/shared/models/Versioning/Blob/CodeBlob';
import { Diff, DiffType } from 'core/shared/models/Versioning/Blob/Diff';
import { exhaustiveCheck } from 'core/shared/utils/exhaustiveCheck';
import {
  convertServerGitBlob,
  convertServerNotebookBlob,
} from '../RepositoryData/Blob/CodeBlob';

export const convertServerCodeDiff = (
  serverCodeDiff: any,
  diffType: DiffType
): ICodeBlobDiff => {
  const category = 'code';
  const { location } = serverCodeDiff;
  if (serverCodeDiff[category].git) {
    switch (diffType) {
      case 'added':
      case 'deleted': {
        const diff: Extract<IGitCodeDiff, { diffType: 'added' | 'deleted' }> = {
          diffType,
          category: 'code',
          type: 'git',
          location,
          blob: {
            category: 'code',
            data: convertServerGitBlob(
              serverCodeDiff[category].git.A || serverCodeDiff[category].git.B
            ),
          },
        } as any;
        return diff;
      }
      case 'updated': {
        const diff: Extract<IGitCodeDiff, { diffType: 'updated' }> = {
          diffType,
          type: 'git',
          category: 'code',
          location,
          blobA: {
            category: 'code',
            data: convertServerGitBlob(serverCodeDiff[category].git.A),
          },
          blobB: {
            category: 'code',
            data: convertServerGitBlob(serverCodeDiff[category].git.B),
          },
        };
        return diff;
      }
      default:
        return exhaustiveCheck(diffType, '');
    }
  }

  if (serverCodeDiff[category].notebook) {
    return convertServerNotebookCodeDiff(
      serverCodeDiff[category].notebook,
      location,
      diffType
    );
  }

  throw new Error('unknown code diff');
};

const convertServerNotebookCodeDiff = (
  serverNotebook: any,
  location: any,
  diffType: DiffType
): INotebookCodeDiff => {
  switch (diffType) {
    case 'added':
    case 'deleted': {
      return {
        type: 'notebook',
        category: 'code',
        diffType: 'added',
        location,
        blob: {
          category: 'code',
          data: convertServerNotebookBlob(serverNotebook.A || serverNotebook.B),
        },
      };
    }
    case 'updated': {
      return {
        type: 'notebook',
        category: 'code',
        diffType: 'updated',
        location,
        blobA: {
          category: 'code',
          data: convertServerNotebookBlob(serverNotebook.A),
        },
        blobB: {
          category: 'code',
          data: convertServerNotebookBlob(serverNotebook.B),
        },
      };
    }
    default:
      return exhaustiveCheck(diffType, '');
  }
};
