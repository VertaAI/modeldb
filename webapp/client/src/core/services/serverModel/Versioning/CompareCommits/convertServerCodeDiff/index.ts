import {
  IGitCodeDiff,
  ICodeBlobDiff,
  INotebookCodeDiff,
} from 'core/shared/models/Versioning/Blob/CodeBlob';
import { exhaustiveCheck } from 'core/shared/utils/exhaustiveCheck';
import { DataLocation } from 'core/shared/models/Versioning/DataLocation';

import {
  convertServerGitBlob,
} from '../../RepositoryData/Blob/CodeBlob';
import { IServerBlobDiff, IServerElementDiff, convertServerElementDiffToClient } from '../ServerDiff';
import { convertServerDatasetPathComponent } from '../../RepositoryData/Blob/DatasetBlob';
import { IServerPathDatasetComponentBlobDiff } from '../convertServerDatasetDiff';

export const convertServerCodeDiff = (
  serverCodeDiff: IServerCodeBlobDiff,
): ICodeBlobDiff => {
  const { location } = serverCodeDiff;
  if (serverCodeDiff.code.git) {
    switch (serverCodeDiff.code.git.status) {
      case 'ADDED': {
        const diff: IGitCodeDiff = {
          diffType: 'added' as const,
          category: 'code',
          type: 'git',
          location: location as DataLocation,
          data: convertServerElementDiffToClient(convertServerGitBlob, serverCodeDiff.code.git),
        };
        return diff;
      }
      case 'DELETED': {
        const diff: IGitCodeDiff = {
          diffType: 'deleted' as const,
          category: 'code',
          type: 'git',
          location: location as DataLocation,
          data: convertServerElementDiffToClient(convertServerGitBlob, serverCodeDiff.code.git),
        };
        return diff;
      }
      case 'MODIFIED': {
        const diff: IGitCodeDiff = {
          diffType: 'updated',
          type: 'git',
          category: 'code',
          location: location as DataLocation,
          data: convertServerElementDiffToClient(convertServerGitBlob, serverCodeDiff.code.git),
        };
        return diff;
      }
      default:
        return exhaustiveCheck(serverCodeDiff.code.git, '');
    }
  }

  if (serverCodeDiff.code.notebook) {
    const serverNotebook = serverCodeDiff.code.notebook;
    switch (serverCodeDiff.status) {
      case 'ADDED': {
        const diff: INotebookCodeDiff = {
          diffType: 'added' as const,
          category: 'code',
          type: 'notebook',
          location: location as DataLocation,
          data: {
            git: serverNotebook.git_repo ? convertServerElementDiffToClient(convertServerGitBlob, serverNotebook.git_repo) : undefined,
            path: serverNotebook.path ? convertServerElementDiffToClient(convertServerDatasetPathComponent, serverNotebook.path) : undefined,
          },
        };
        return diff;
      }
      case 'DELETED': {
        const diff: INotebookCodeDiff = {
          diffType: 'deleted' as const,
          category: 'code',
          type: 'notebook',
          location: location as DataLocation,
          data: {
            git: serverNotebook.git_repo ? convertServerElementDiffToClient(convertServerGitBlob, serverNotebook.git_repo) : undefined,
            path: serverNotebook.path ? convertServerElementDiffToClient(convertServerDatasetPathComponent, serverNotebook.path) : undefined,
          },
        };
        return diff;
      }
      case 'MODIFIED': {
        const diff: INotebookCodeDiff = {
          diffType: 'updated' as const,
          category: 'code',
          type: 'notebook',
          location: location as DataLocation,
          data: {
            git: serverNotebook.git_repo ? convertServerElementDiffToClient(convertServerGitBlob, serverNotebook.git_repo) : undefined,
            path: serverNotebook.path ? convertServerElementDiffToClient(convertServerDatasetPathComponent, serverNotebook.path) : undefined,
          },
        };
        return diff;
      }
      default:
        return exhaustiveCheck(serverCodeDiff.status, '');
    }
  }

  throw new Error('unknown code diff');
};

export type IServerCodeBlobDiff = IServerBlobDiff<{
  code: {
    git?: IServerGitDiff;
    notebook?: {
      path?: IServerPathDatasetComponentBlobDiff;
      git_repo?: IServerGitDiff;
    };
  }
}>;

type IServerGitDiff = IServerElementDiff<{
  repo: string;
  hash?: string;
  branch?: string;
  tag?: string;
  is_dirty: boolean;
}>;
