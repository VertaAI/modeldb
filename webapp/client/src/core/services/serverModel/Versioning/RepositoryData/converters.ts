import {
  IBlobFolderElement,
  IFolder,
  ISubFolderElement,
  ICommit,
} from 'core/shared/models/Versioning/RepositoryData';

export const convertServerFolderToClient = (serverFolder: any): IFolder => {
  return {
    type: 'folder',
    blobs: (serverFolder.blobs || []).map(
      ({ element_name, created_by_commit }: any) => {
        const res: IBlobFolderElement = {
          type: 'blob',
          name: element_name,
          createdByCommitSha: created_by_commit,
        };
        return res;
      }
    ),
    subFolders: (serverFolder.sub_folders || []).map(
      ({ element_name, created_by_commit }: any) => {
        const res: ISubFolderElement = {
          type: 'folder',
          name: element_name,
          createdByCommitSha: created_by_commit,
        };
        return res;
      }
    ),
  };
};

export const convertServerCommitToClient = (serverCommit: any): ICommit => {
  return {
    authorId: serverCommit.author,
    dateCreated: new Date(Number(serverCommit.date_created)),
    sha: serverCommit.commit_sha,
    message: serverCommit.message,
    parentShas: serverCommit.parent_shas || [],
    type: serverCommit.parent_shas ? 'withParent' : 'initial',
  };
};
