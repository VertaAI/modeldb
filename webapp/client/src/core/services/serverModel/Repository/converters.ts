import { IRepository, Label } from 'core/shared/models/Repository/Repository';
import { convertServerEntityWithLoggedDates } from 'services/serverModel/Common/converters';
import { convertServerShortWorkspaceToClient } from 'services/serverModel/Workspace/converters';

import User from 'models/User';
import { IServerRepository } from './Repository';

export const convertServerRepositoryToClient = ({
  serverRepository,
  labels,
  owner,
}: {
  serverRepository: IServerRepository;
  labels: Label[];
  owner: User;
}): IRepository => {
  const { dateCreated, dateUpdated } = convertServerEntityWithLoggedDates({
    date_created: serverRepository.date_created,
    date_updated: serverRepository.date_updated,
  });
  return {
    id: serverRepository.id,
    name: serverRepository.name,
    shortWorkspace: convertServerShortWorkspaceToClient(serverRepository),
    dateCreated,
    dateUpdated,
    labels,
    owner,
  };
};
