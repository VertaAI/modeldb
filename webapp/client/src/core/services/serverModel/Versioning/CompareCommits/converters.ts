import * as R from 'ramda';

import { IBlob, blobCategories } from 'core/shared/models/Repository/Blob/Blob';
import { Diff, DiffType } from 'core/shared/models/Repository/Blob/Diff';
import { exhaustiveCheck } from 'core/shared/utils/exhaustiveCheck';
import matchType from 'core/shared/utils/matchType';

import { convertServerCodeDiff } from './convertServerCodeDiff';
import { convertServerConfigDiff } from './convertServerConfigDiff';
import convertServerDatasetDiff from './convertServerDatasetDiff';
import { convertServerEnvironmentDiff } from './convertServerEnvironmentDiff';

export const convertServerDiffsToClient = (serverDiff: any[]): Diff[] => {
  return (serverDiff || [])
    .filter((diff: any) => !R.isEmpty(diff))
    .map(convertServerDiffToClient);
};

export const convertServerDiffToClient = (serverDiff: any): Diff => {
  const serverDiffCategory = Object.keys(
    R.omit(['location', 'status'], serverDiff)
  )[0];
  const category: IBlob['data']['category'] =
    serverDiffCategory in blobCategories
      ? (serverDiffCategory as IBlob['data']['category'])
      : 'unknown';

  const diffType: DiffType = serverDiff.status
    ? matchType(
        {
          MODIFIED: () => 'updated',
          DELETED: () => 'deleted',
          ADDED: () => 'added',
        },
        serverDiff.status
      )
    : 'added';

  switch (category) {
    case 'code': {
      return convertServerCodeDiff(serverDiff, diffType);
    }

    case 'config': {
      return convertServerConfigDiff(serverDiff, diffType);
    }

    case 'environment': {
      return convertServerEnvironmentDiff(serverDiff, diffType);
    }

    case 'unknown': {
      switch (diffType) {
        case 'added':
        case 'deleted':
          return {
            diffType,
            category: 'unknown',
            type: 'unknown',
            location: serverDiff.location,
            blob: serverDiff[serverDiffCategory],
          };

        case 'updated': {
          return {
            diffType,
            category: 'unknown',
            type: 'unknown',
            location: serverDiff.location,
            blobA: serverDiff[serverDiffCategory].A,
            blobB: serverDiff[serverDiffCategory].B,
          };
        }

        default:
          exhaustiveCheck(diffType, '');
      }
    }

    case 'dataset': {
      return convertServerDatasetDiff(serverDiff, diffType);
    }

    default: {
      exhaustiveCheck(category, '');
    }
  }
};
