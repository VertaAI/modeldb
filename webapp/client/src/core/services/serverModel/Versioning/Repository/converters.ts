import { IRepository, Label } from 'core/shared/models/Versioning/Repository';
import { convertServerEntityWithLoggedDates } from 'services/serverModel/Common/converters';
import { convertServerShortWorkspaceToClient } from 'services/serverModel/Workspace/converters';

import { IServerRepository } from './Repository';
import matchType from 'core/shared/utils/matchType';
import User from 'models/User';

export const convertServerRepositoryToClient = ({
  serverRepository,
  labels,
  owner,
}: {
  serverRepository: IServerRepository;
  labels: Label[];
  owner: User;
}): IRepository => {
  const { dateCreated, dateUpdated } = convertServerEntityWithLoggedDates({
    date_created: serverRepository.date_created,
    date_updated: serverRepository.date_updated,
  });
  return {
    id: serverRepository.id,
    name: serverRepository.name,
    shortWorkspace: convertServerShortWorkspaceToClient(serverRepository),
    dateCreated,
    dateUpdated,
    labels,
    owner,
    visibility: serverRepository.repository_visibility
      ? matchType<
          Required<IServerRepository>['repository_visibility'],
          IRepository['visibility']
        >(
          {
            ORG_SCOPED_PUBLIC: () => 'organizationPublic',
            PRIVATE: () => 'private',
            PUBLIC: () => 'public',
          },
          serverRepository.repository_visibility
        )
      : 'private',
  }
};
