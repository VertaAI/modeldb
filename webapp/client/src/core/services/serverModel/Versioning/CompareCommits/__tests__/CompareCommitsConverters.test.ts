import { convertServerDiffToClient } from '../converters';
import { convertServerCodeDiff } from '../convertServerCodeDiff';
import {
  mockedClientSuccessAddedDatasetDiff,
  mockedServerAddedDatasetDiff,
  mockedServerDeletedDatasetDiff,
  mockedServerUpdatedDatasetDiff,
  mockedClientSuccessDeletedDatasetDiff,
  mockedClientSuccessUpdatedDatasetDiff,
} from './mocks/mockCompareCommits';

describe('CompareCommits converters', () => {
  describe('(convertServerDiffToClient)', () => {
    describe('dataset blob diff', () => {
      it('should convert correct server response with status "added" to client data', () => {
        expect(convertServerDiffToClient(mockedServerAddedDatasetDiff)).toEqual(
          mockedClientSuccessAddedDatasetDiff
        );
      });

      it('should convert correct server response with status "deleted" to client data', () => {
        expect(
          convertServerDiffToClient(mockedServerDeletedDatasetDiff)
        ).toEqual(mockedClientSuccessDeletedDatasetDiff);
      });

      it('should convert correct server response with status "updated" to client data', () => {
        expect(
          convertServerDiffToClient(mockedServerUpdatedDatasetDiff)
        ).toEqual(mockedClientSuccessUpdatedDatasetDiff);
      });
    });

    describe('code blob diff', () => {
      it('should convert git code diff with deleted diff type', () => {
        const diff = {
          location: ['deleted'],
          status: 'DELETED',
          code: {
            git: {
              B: {
                repo: 'git@github.com:VertaAI/modeldb.git',
                hash: '#hhh',
                branch: 'development',
                tag: 'tag-2',
              },
            },
          },
        };

        expect(convertServerDiffToClient(diff)).toMatchSnapshot();
      });

      it('should convert git code diff with modified diff type', () => {
        const diff = {
          location: ['folder_asdfc', 'repo_blob'],
          status: 'MODIFIED',
          code: {
            git: {
              A: {
                repo: 'git@github.com:VertaAI/modeldb.git',
                hash: '#hhh',
                branch: 'master',
                tag: 'tag',
                is_dirty: true,
              },
              B: {
                repo: 'git@github.com:VertaAI/modeldb.git',
                hash: '#hhh',
                branch: 'development',
                tag: 'tag-2',
              },
            },
          },
        };

        expect(convertServerDiffToClient(diff)).toMatchSnapshot();
      });
    });
  });
});
