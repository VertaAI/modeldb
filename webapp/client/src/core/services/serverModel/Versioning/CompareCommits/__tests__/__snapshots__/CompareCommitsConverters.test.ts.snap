// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompareCommits converters (convertServerDiffToClient) code blob diff should convert git code diff with deleted diff type 1`] = `
Object {
  "blob": Object {
    "category": "code",
    "data": Object {
      "data": Object {
        "branch": "development",
        "commitHash": "#hhh",
        "isDirty": undefined,
        "remoteRepoUrl": Object {
          "type": "github",
          "value": "git@github.com:VertaAI/modeldb.git",
        },
        "tag": "tag-2",
      },
      "type": "git",
    },
  },
  "category": "code",
  "diffType": "deleted",
  "location": Array [
    "deleted",
  ],
  "type": "git",
}
`;

exports[`CompareCommits converters (convertServerDiffToClient) code blob diff should convert git code diff with modified diff type 1`] = `
Object {
  "blobA": Object {
    "category": "code",
    "data": Object {
      "data": Object {
        "branch": "master",
        "commitHash": "#hhh",
        "isDirty": true,
        "remoteRepoUrl": Object {
          "type": "github",
          "value": "git@github.com:VertaAI/modeldb.git",
        },
        "tag": "tag",
      },
      "type": "git",
    },
  },
  "blobB": Object {
    "category": "code",
    "data": Object {
      "data": Object {
        "branch": "development",
        "commitHash": "#hhh",
        "isDirty": undefined,
        "remoteRepoUrl": Object {
          "type": "github",
          "value": "git@github.com:VertaAI/modeldb.git",
        },
        "tag": "tag-2",
      },
      "type": "git",
    },
  },
  "category": "code",
  "diffType": "updated",
  "location": Array [
    "folder_asdfc",
    "repo_blob",
  ],
  "type": "git",
}
`;
